File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


func invertTree(root *TreeNode) *TreeNode {
    if root == nil {
        return nil
    }

    someNode := root.Left
    root.Left = root.Right
    root.Right = someNode

    invertTree(root.Left)
    invertTree(root.Right)

    return root

}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int idx = 0;
    HashMap<Integer, Integer> mpp = new HashMap<>();
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i = 0; i < inorder.length; i++) {
            mpp.put(inorder[i], i);
        }

        return dfs(preorder, 0, inorder.length - 1);
    }
    public TreeNode dfs(int[] pre, int l, int r) {
        if (l > r)
            return null;
        int val = pre[idx++];
        TreeNode root = new TreeNode(val);
        int m = mpp.get(val);
        root.left = dfs(pre, l, m - 1);
        root.right = dfs(pre, m + 1, r);
        return root;
    }
}
